Refresh access_token

Naokimi opened the pull request:
Bolt has developed an option to refresh the access_token after
expiration. In order to do that we have to send an API call to the same
endpoint with a different payload. This commit implements that option.

Requirement for https://github.com/nebulab/solidus_bolt/issues/101

Manually tested: `refresh_token`, `refresh_token_scope`, and `expiration_time` do appear in the session

Naokimi committed:
Pass variables to refresh access_token

Bolt has developed an option to refresh the access_token after
expiration. In order to do that we have to send an API call to the same
endpoint with a different payload. This commit passes the variables
necessary to make the call.

DanielePalombo commented on lib/omniauth/strategies/bolt.rb:
```diff
@@ -62,6 +74,9 @@ def callback_phase
         response = JSON.parse(res.body)
         @access_token = response['access_token']
         @user_uid = response['id_token']
+        session[:expiration_time] = Time.now + response['expires_in']
+        session[:refresh_token] = response['refresh_token']
+        session[:refresh_token_scope] = response['refresh_token_scope']
```
Can you add the `bolt_` prefix to these fields, and the `authorization_code` as well?

Naokimi replied:
reworked logic in https://github.com/nebulab/omniauth-bolt/pull/2/commits/9b3154e439f4f511d1c1335517c5b9ba666567ab

DanielePalombo commented on lib/omniauth/strategies/bolt.rb:
```diff
@@ -39,14 +39,26 @@ def request_phase
           client_secret: options['api_key']
         }
 
+        token_and_scope = if session['refresh_token']
+                            {
+                              grant_type: 'refresh_token',
+                              refresh_token: session['refresh_token'],
+                              scope: session['refresh_token_scope']
+                            }
+                          else
+                            {
+                              grant_type: 'authorization_code',
+                              code: session['authorization_code'],
+                              scope: session['scope']
+                            }
+                          end
+
+        payload.merge!(token_and_scope)
```
I don't understand the purpose of this change.
Here the user is logged in, and a new `session['authorization_code']` is provided [here](https://github.com/nebulab/omniauth-bolt/pull/2/commits/acdcdc3b08cb534d9c94a60125ae6c7aa3c46b99#diff-c0c57b12123a23b6cbe91332a48b0a8eda6ed6ed3c454f6e0e279445cd79f437L34), so the access_token will be never expired here.

Naokimi replied:
reworked logic in https://github.com/nebulab/omniauth-bolt/pull/2/commits/9b3154e439f4f511d1c1335517c5b9ba666567ab

DanielePalombo commented on lib/omniauth/strategies/bolt.rb:
```diff
@@ -39,14 +39,26 @@ def request_phase
+                          else
+                            {
+                              grant_type: 'authorization_code',
+                              code: session['authorization_code'],
+                              scope: session['scope']
```
From now these session variables won't be used anymore. 
What do you think to delete them from the session?

Naokimi replied:
reworked logic in https://github.com/nebulab/omniauth-bolt/pull/2/commits/9b3154e439f4f511d1c1335517c5b9ba666567ab

DanielePalombo replied:
You marked this comment as resolved, but it doesn't look resolved to me 🤔

Naokimi replied:
fixed in https://github.com/nebulab/omniauth-bolt/pull/2/commits/986cd543eb3c2d65eec5b70f9092820715ee2c58

DanielePalombo reviewed (commented):
The access_token was not set to the session but just assigned to an instance variable, to let the developer to implement its own logic and decide where to store it(DB, session, Redis, etc...).
What do you think to keep the same logic here for refresh_token and others?

DanielePalombo commented on lib/omniauth/strategies/bolt.rb:
```diff
@@ -60,6 +60,9 @@ def callback_phase
         post_data = URI.encode_www_form(payload)
         res = http.request(req, post_data)
         response = JSON.parse(res.body)
+        @expiration_time = Time.now.utc + 15.minutes
```
Why `15.minutes`? The expiration is defined by Bolt.
There is a `response['expires_in']` which defines the expiration time, it should be extended to the sandbox environment to 1 hour.
The expiration_time should be defined as `Time.current + response['expires_in'].seconds`

Naokimi replied:
fixed in https://github.com/nebulab/omniauth-bolt/pull/2/commits/56a86a997df1eccdacb8a76cf88d77482795dfcc

DanielePalombo commented on lib/omniauth/strategies/bolt.rb:
```diff
@@ -60,7 +60,7 @@ def callback_phase
         res = http.request(req, post_data)
         response = JSON.parse(res.body)
-        @expiration_time = Time.now.utc + 15.minutes
+        @expiration_time = Time.now.utc + response['expires_in']
```
Does it assume that these are seconds?

Naokimi replied:
yes - it's 3600 which gets automatically interpreted as seconds

Naokimi replied:
the payload is 3600
```
[2] pry(main)> Time.now.utc + 3600
=> 2022-06-03 09:17:09.6563547 UTC
[3] pry(main)> Time.now.utc + 3600.seconds
=> 2022-06-03 09:17:12.8559983 UTC
```
also in pure Ruby we can't call `.seconds` on an integer

Naokimi committed:
Clear session after generating access_token

Since authorization_code and scope aren't being used when refreshing
the access_token, keeping them in the session is redundat. That's why
we'll remove them after generating the code for the first time.
